File: code/7._Lempel-Ziv-Welch(LZW)/Python/programming.py/LZW.py
def compress(input_string):
    dictionary = {chr(i): i for i in range(256)}
    result = []
    start = 0
    current_code = 256

    for i in range(1, len(input_string) + 1):
        current_substring = input_string[start:i]

        if current_substring not in dictionary and i < len(input_string):
            dictionary[current_substring] = current_code
            current_code += 1
        else:
            if i == len(input_string):
                result.append(dictionary[current_substring])
            start = i

    return result


def decompress(compressed_data):
    dictionary = {i: chr(i) for i in range(256)}
    result = [chr(compressed_data[0])]
    current_code = 256

    for code in compressed_data[1:]:
        if code in dictionary:
            entry = dictionary[code]
        elif code == current_code:
            entry = result[0]
        else:
            raise ValueError("Corrupted data")

        result.append(entry)
        dictionary[current_code] = result[0]
        current_code += 1

    return "".join(result)


# Example usage:
original_text = "ABABABABA"
compressed_data = compress(original_text)
print("Compressed Data:", compressed_data)
decompressed_text = decompress(compressed_data)
print("Decompressed Text:", decompressed_text)

File: code/7._Lempel-Ziv-Welch(LZW)/Java/programming.java/LZW.java
import java.util.*;

public class LZWCompression {
    public static List<Integer> compress(String input) {
        Map<String, Integer> dictionary = new HashMap<>();
        for (int i = 0; i < 256; i++) {
            dictionary.put(Character.toString((char) i), i);
        }

        List<Integer> result = new ArrayList<>();
        int start = 0;
        int currentCode = 256;

        for (int i = 1; i <= input.length(); i++) {
            String currentSubstring = input.substring(start, i);

            if (!dictionary.containsKey(currentSubstring) && i < input.length()) {
                dictionary.put(currentSubstring, currentCode);
                currentCode++;
            } else {
                if (i == input.length()) {
                    result.add(dictionary.get(currentSubstring));
                }
                start = i;
            }
        }

        return result;
    }

    public static String decompress(List<Integer> compressedData) {
        Map<Integer, String> dictionary = new HashMap<>();
        for (int i = 0; i < 256; i++) {
            dictionary.put(i, Character.toString((char) i));
        }

        List<String> result = new ArrayList<>();
        result.add(dictionary.get(compressedData.get(0)));
        int currentCode = 256;

        for (int i = 1; i < compressedData.size(); i++) {
            int code = compressedData.get(i);
            String entry;

            if (dictionary.containsKey(code)) {
                entry = dictionary.get(code);
            } else if (code == currentCode) {
                entry = result.get(0);
            } else {
                throw new IllegalArgumentException("Corrupted data");
            }

            result.add(entry);
            dictionary.put(currentCode, result.get(0));
            currentCode++;
        }

        StringBuilder sb = new StringBuilder();
        for (String s : result) {
            sb.append(s);
        }

        return sb.toString();
    }

    public static void main(String[] args) {
        String originalText = "ABABABABA";
        List<Integer> compressedData = compress(originalText);
        System.out.println("Compressed Data: " + compressedData);
        String decompressedText = decompress(compressedData);
        System.out.println("Decompressed Text: " + decompressedText);
    }
}
